src/types.rs

```diff
@@ -2,11 +2,103 @@
 
 use clap::{Parser, ValueEnum};
 use serde::Serialize;
-use std::time::Duration;
+use std::{str::FromStr, time::Duration};
+use streamstore::types::{BasinName, ConvertError};
 
 pub const STORAGE_CLASS_PATH: &str = "default_stream_config.storage_class";
 pub const RETENTION_POLICY_PATH: &str = "default_stream_config.retention_policy";
 
+#[derive(Debug, Clone)]
+pub struct BasinNameOrUri<S> {
+    pub basin: BasinName,
+    pub stream: S,
+}
+
+impl<S> From<BasinNameOrUri<S>> for BasinName {
+    fn from(value: BasinNameOrUri<S>) -> Self {
+        value.basin
+    }
+}
+
+fn parse_maybe_basin_or_uri(s: &str) -> Result<(BasinName, Option<String>), ConvertError> {
+    match BasinName::from_str(s) {
+        Ok(basin) => {
+            // Definitely a basin name since a valid basin name cannot have `:`
+            // which is required for the URI.
+            Ok((basin, None))
+        }
+        Err(parse_basin_err) => {
+            // Should definitely be a URI else error.
+            let uri = http::Uri::from_str(s).map_err(|_| parse_basin_err)?;
+
+            match uri.scheme_str() {
+                Some("s2") => (),
+                Some(other) => return Err(format!("Invalid S2 URL scheme: '{other}'").into()),
+                None => return Err("S2 URL scheme empty".into()),
+            };
+
+            let basin = uri.host().ok_or("Basin name missing in S2 URL")?;
+            let basin = BasinName::from_str(basin)
+                .map_err(|e| format!("Invalid basin name in S2 URL: {e}"))?;
+
+            let stream = uri.path().trim_start_matches('/');
+            let stream = if stream.is_empty() {
+                None
+            } else {
+                Some(stream.to_string())
+            };
+
+            Ok((basin, stream))
+        }
+    }
+}
+
+pub type BasinNameOnlyUri = BasinNameOrUri<()>;
+
+impl FromStr for BasinNameOnlyUri {
+    type Err = ConvertError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (basin, stream) = parse_maybe_basin_or_uri(s)?;
+        if stream.is_none() {
+            Ok(Self { basin, stream: () })
+        } else {
+            Err("Expected S2 URL with only basin name".into())
```

Comment:
Can we instead return a CLI error and have a help section for it? Would that be feasible?

Reply:
So, the help section here is useless since this is parsed by `clap` which doesn't care about `miette::Diagnostic`.

I have made this change though, so let's stick with it, maybe helps if the code changes in future?